### PR TITLE
Use language server implementation instead of language for URLs

### LINKED ISSUE #192: Use implementation name rather than language for WebSocket URLs
## Elevator Pitch

Use a more precise definition of _which_ language server should be used for a connection in URLs.

## Motivation

At present, the WebSocket URL scheme is rather nebulous in that _all_ language servers that register for a language will be connected during a `Session`, which is probably not right and would no doubt explode if there were multiple providers for a given language. Further the "language" concept doesn't _really_ provide enough depth. Combined with #190, we should have enough information to do a much better job of detecting (or deferring to the user) which implementation to use in a particular case.

Additionally, we can save extra connections when multiple documents of different languages connect to the same polyglot server, rather than hiding it.

## Design Ideas

While we're at it, let's add another (short) prefix to the WebSocket URLs so that we are less constrained in potential future endpoints.

| Old | New |
|-|-|
| `/lsp` | `/lsp/status` |
| `/lsp/python`     | `/lsp/ws/pyls` |
| `/lsp/javascript`<br/>`/lsp/jsx`<br/>`/lsp/typescript`<br/>`/lsp/typescript-jsx`<br/>`/lsp/typescriptreact`<br/>`/lsp/javascriptreact` | `/lsp/ws/javascript-typescript-langserver` |

### PR DESCRIPTION
**Pull request by bollwyvl**
## References

- fixes #192

## Code changes

- backend
  - [x] uses implementation name for websocket key, e.g. `/lsp/ws/pyls`
  - [x] moves status to `/lsp/status`
  - [x] move reading/writing to stdio language servers to threads
  - [x] move reading/writing to shadow files to threads

- [x] fix frontend
  - [x] introduce `ILanguageServerManager`
  - [x] use `ILSM` in `ConnectionManager` and `StatusBarItem`

## User-facing changes

- None, other than launching a document with a language already supported by another language will be faster, as there's only one actual WebSocket/process behind them

## Backwards-incompatible changes

- [x] bumps schema to version `2`
  - [x] changes session payload to be a dict, e.g. `{"session": {"pyls": {}}}`

## Chores

- [x] linted
- [x] tested
- [x] documented
- [x] changelog entry

### CODE REVIEW

**Comment by bollwyvl** [comment]
starting point for getting the frontend to compile:
```
../jupyterlab-lsp/src/adapters/jupyterlab/components/statusbar.tsx(341,7): error TS2740: Type '{ [k: string]: LanguageServerSession; }' is missing the following properties from type 'LanguageServerSession[]': length, pop, push, concat, and 26 more.
```
here we go...

File: py_src/jupyter_lsp/tests/conftest.py [hunk 1]
```diff
@@ -12,45 +12,39 @@
...
-    "json",
-    "jsx",
-    "less",
-    "markdown",
-    "python",
-    "scss",
-    "typescript-jsx",
-    "typescript",
-    "yaml",
+KNOWN_SERVERS = [
```
Comment (bollwyvl) [inline py_src/jupyter_lsp/tests/conftest.py]: nice thing here is it is now slightly faster...

**Comment by bollwyvl** [comment]
Turns out I didn't _have_ to do #190, so I didn't. Otherwise seems pretty good!

File: atest/01_Editor.robot [hunk 2]
```diff
@@ -19,7 +19,7 @@ CSS
 
 Docker
     ${def} =    Set Variable    xpath://span[contains(@class, 'cm-string')][contains(text(), 'PLANET')]
-    Editor Shows Features for Language    Docker    Dockerfile    Diagnostics=Instruction has no arguments    Jump to Definition=${def}    Rename=${def}
+    Wait Until Keyword Succeeds  3x  100ms  Editor Shows Features for Language    Docker    Dockerfile    Diagnostics=Instruction has no arguments    Jump to Definition=${def}    Rename=${def}
```
Comment (bollwyvl) [inline atest/01_Editor.robot]: So this "fixes" the problem, but the fact remains that you don't reliably get diagnostics back from a dockerfile on the first file opened until you change the file (then it seems to work every time). I tried sending more change events, but it didn't have an effect. I'm willing to deal with it if it makes the tests pass for now.

File: py_src/jupyter_lsp/stdio.py [hunk 3]
```diff
@@ -34,6 +40,7 @@ def __repr__(self):  # pragma: no cover
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
         self.log.debug("%s initialized", self)
+        self.executor = ThreadPoolExecutor(max_workers=1)
```
Comment (bollwyvl) [inline py_src/jupyter_lsp/stdio.py]: I thought all of the jiggery pokery i did with the streams was sufficient to handle this, but I've still hit some slowdown. This should Make Sure we don't block the main thread on IO below

File: py_src/jupyter_lsp/stdio.py [hunk 4]
```diff
@@ -87,7 +94,7 @@ def wake(self):
                 else:
                     self.wake()
 
-                await self.queue.put(message)
+                IOLoop.current().add_callback(self.queue.put_nowait, message)
```
Comment (bollwyvl) [inline py_src/jupyter_lsp/stdio.py]: again, I don't _think_ this was causing issues, but better be safe...

File: py_src/jupyter_lsp/stdio.py [hunk 5]
```diff
@@ -98,12 +105,12 @@ def wake(self):
         message = ""
         headers = HTTPHeaders()
 
-        line = self._readline()
+        line = await convert_yielded(self._readline())
```
Comment (bollwyvl) [inline py_src/jupyter_lsp/stdio.py]: there are other wrappers that wouldn't require converting, but this works...

File: py_src/jupyter_lsp/stdio.py [hunk 6]
```diff
@@ -142,17 +150,21 @@ class LspStdIoWriter(LspStdIoBase):
...
-                self.stream.flush()
+                await convert_yielded(self._write_one(response.encode("utf-8")))
             except Exception:  # pragma: no cover
                 self.log.exception("s couldn't write message: %s", self, response)
             finally:
                 self.queue.task_done()
+
+    @run_on_executor
+    def _write_one(self, message) -> None:
+        self.stream.write(message)
```
Comment (bollwyvl) [inline py_src/jupyter_lsp/stdio.py]: could have moved more string operations here, but keeps the critical region more clear

File: packages/jupyterlab-lsp/src/virtual/editor.spec.ts [hunk 7]
```diff
@@ -59,6 +60,17 @@ describe('VirtualEditor', () => {
...
   );
 
+  const LANGSERVER_MANAGER = new MockLanguageServerManager({});
+  const CONNECTION_MANAGER = new DocumentConnectionManager({
+    language_server_manager: LANGSERVER_MANAGER
+  });
+
+  const DEBUG = false;
+
+  if (DEBUG) {
```
Comment (krassowski) [inline packages/jupyterlab-lsp/src/virtual/editor.spec.ts]: a debug - left intentionally or just slipped in?

Comment (bollwyvl) [reply packages/jupyterlab-lsp/src/virtual/editor.spec.ts]: yeah, accidental!

**Comment by bollwyvl** [comment]
I'll have a look at fixing the merge conflicts presently...

**Comment by krassowski** [comment]
Ok, I will wait!

File: packages/jupyterlab-lsp/src/virtual/editor.spec.ts (hunk 7 above)
Comment (bollwyvl) [reply packages/jupyterlab-lsp/src/virtual/editor.spec.ts]: Nope, needs it! Have changed to make it part of the actual test environment.

**Comment by bollwyvl** [comment]
Ok, provided docker doesn't flake out on us, this is good to go if CI passes.

File: py_src/jupyter_lsp/virtual_documents_shadow.py [hunk 8]
```diff
@@ -16,11 +23,19 @@ def extract_or_none(obj, path):
 
 
 class EditableFile:
+    executor = ThreadPoolExecutor(max_workers=MAX_WORKERS)
```
Comment (bollwyvl) [inline py_src/jupyter_lsp/virtual_documents_shadow.py]: following the same approach as stdio, but in this case we have one executor for all instances of `EditableFile`. 4 seems like enough for now.

File: py_src/jupyter_lsp/virtual_documents_shadow.py [hunk 9]
```diff
@@ -16,11 +23,19 @@ def extract_or_none(obj, path):
...
     def __init__(self, path):
         # Python 3.5 relict:
         self.path = Path(path) if isinstance(path, str) else path
-        self.lines = self.read_lines()
 
+    async def read(self):
+        self.lines = await convert_yielded(self.read_lines())
+
+    async def write(self):
+        return await convert_yielded(self.write_lines())
```
Comment (bollwyvl) [inline py_src/jupyter_lsp/virtual_documents_shadow.py]: these just wrap the underlying tornado yield-style async stuff.

File: py_src/jupyter_lsp/virtual_documents_shadow.py [hunk 10]
```diff
@@ -125,7 +141,9 @@ def setup_shadow_filesystem(virtual_documents_uri):
             return
 
         path = file_uri_to_path(uri)
-        file = EditableFile(path)
+        editable_file = EditableFile(path)
+
+        await editable_file.read()
```
Comment (bollwyvl) [inline py_src/jupyter_lsp/virtual_documents_shadow.py]: renaming is cosmetic, just don't need to go re-defining builtins without a good reason.

It's a little ugly to have to do the read: could add an `async` class method to `EditableFile` to do both in one go...

**Comment by bollwyvl** [comment]
have definitely still been getting the dockerfile issue locally... very challenging to reproduce, apparently...

File: py_src/jupyter_lsp/virtual_documents_shadow.py (hunk 10 above)
Comment (krassowski) [reply py_src/jupyter_lsp/virtual_documents_shadow.py]: It's a small hill, but [file is not a builtin in Python3](https://stackoverflow.com/a/24942363/6646912) and it is ok to use it. It was in Python2, and yes some syntax highlighters did not catch up, but we can use it now.

**Comment by bollwyvl** [comment]
ha, fair.

**Comment by bollwyvl** [comment]
I swear I'm done tinkering now.

**Comment by krassowski** [comment]
Now let's hope we will have CI pass on master ;)

**Comment by bollwyvl** [comment]
Hooray merge! Boo ci fail! We could mark it as non-critical, but those have
a tendency to get ignored. looks like there's a new version we could try. I
guess we need an issue.

Aside from fixing the docker issue, what's the next thing we want to get
to? With this PR in, per-server configuration is unblocked, and would help
a lot, as is the mimetype-based detection stuff.

I'm also interested in the multiple connection providers: being able to
install js-only servers as lazy loading labextensions would help the
install complexity around run-time node without negatively impacting
performance much: and json and markdown would be great initial candidates,
especially if more things like settings manager/cell metadata and markdown
cells could be extended.

>

**Pull request merged**